README.md
SubNetwork Explorer v0.0.3BETA1-test1-prerelease
--------------------------------------------------------------------------------
Copyright (C) 2000 Prana 
License: GPLv2

** Note: This tiny utility is now archived and no longer developed. Uploaded to
Github on Aug 20, 2016. 

Description
================================================================================
"SubNetwork Explorer" is a program that scans the subnets of a network
for anonymous Ftp, Netbios, and SunRPC. If no options specified, it will just
display the hostnames without scanning them. Uses 'fork()' to quickly scan the 
subnets all at the same time. This program will check for ftp and tries to login
to it anonymously.

For example:

$ nslookup localhost
Server:  localhost
Address:  127.0.0.1

Non-authoritative answer:
Name:    localhost
Address:  127.0.0.1

$ netxplorer 127.0.*.*
master.localhost
host1.localhost
host2.localhost
host3.localhost

Requirements
================================================================================
* Linux, FreeBSD, Solaris, or any other UN*X
* G++ (GNU C++) compiler
* glibc-devel


Installation
================================================================================

* Just type "make".
* If you feel that the executable size is too big (300k), just strip the
  debug information by typing "strip netxplorer"
* If you want to put it in /usr/bin, type "make install" 
  (you have to login as root) for "make install"


Usage
================================================================================
netxplorer [options] [host range]
 OPTIONS:
 -? : show this help screen
 -c : perform a connect() port scanning
 -f : perform a half connection scan - undetected by Odin's iplog
 -a : scan all port from 1-65535
 -p : only scan privileged port from 1-1024
 -n : no scan - only query hostnames (DEFAULT)
 -s : scan specific port only:
      -s cups	 : scan CUPS printing server
      -s ftp	 : scan ftp server
      -s http 	 : scan http server
      -s mail 	 : scan mail server
      -s lpd	 : scan lpd printing server
      -s netbios : scan netbios
      -s sunrpc  : scan sunrpc
      -s ssh 	 : scan ssh
 -v : try to retrieve versions/informations from port
 [host range] : the network you want to scan

 EXAMPLE:
 netxplorer 207.168.230.* will display the hostnames from 207.168.230.0-255 *WITHOUT* scanning
 netxplorer localhost -s http -s cups will scan 'localhost' for http and CUPS server
 netxplorer -c -a 192.168.1.* will scan from 192.168.1.0-255 and perform connect() scan from port 1-65535

 NOTE: Host *MUST* be in number format (eg: 127.0.*.*) if you want to do multiple host
 scanning with the '*' asterisk .
 This program will then do a forking for each hosts. Without a parameter,
 the result will only display the host names

FAQ
================================================================================
Q: This program is so buggy, I don't like it!!
A: Please note that it's still a BETA version, so it *IS* buggy.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Q: Why isn't it capable of TCP SYN Scan like nmap?
A: Later in v0.0.3BETA
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Q: Why does it only scans for ftp, netbios, and sunrpc, cups, etc?
A: I will add more for feature release, including Proxy & Wingates...
   The code is available, but I want to make more testing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Q: Do you have the RPM package of the program?
A: Currently no. I will pack it in RPM after it's not that buggy.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


ChangeLog
Jan 1, 2000: v0.0.3BETA1-test1-prerelease
* Fixed connect_scanner::scan_port(...) 
* Fixed memory hog
* Added half-connection scan - not detected by Odin's iplog ;) hehehe
* Added: web, mail, CUPS server scans.

December 22, 2000: v0.0.2BETA
* Rewrite everything with C++
* Improved code
* Major code cleanup
* Single host scanning
* More complete Sun RPC
* Added: web, mail, CUPS server scans.

July 19, 2000: v0.0.1BETA
* Initial public release

July 16, 2000:
* Project started


scanner.h
// scanner.h
/*****************************************************************************
 *  SubNetwork Explorer                                       	             *
 *  Copyright (C) 2000 Prana                              *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/
#ifndef SCANNER_H
#define SCANNER_H
 
#include <string>

#define MAX_SCANNER_CONFIG		16

#define SCANNER_READY			1
#define SCANNER_OPT_SCAN_ALL		2
#define SCANNER_OPT_SCAN_1024		3
#define SCANNER_OPT_NOSCAN		4

// modules
#define SCANNER_M_HTTP			5
#define SCANNER_M_FTP			6
#define SCANNER_M_SUNRPC		7
#define SCANNER_M_CUPS			8
#define SCANNER_M_NETBIOS		9
#define SCANNER_M_MAIL			10
#define SCANNER_M_LPD			11
#define SCANNER_M_SSH			12
typedef bool (*scanner_port_modules) (int);

class Scanner
{
   public:
      Scanner();
      ~Scanner();
      virtual bool set_hostname(const char *name);
      string get_hostname();
      bool set_arguments(string *args);
      bool ping();
      virtual bool scan_port(int port) = 0;
      void scan();
      bool set_time_out(int descriptor, int sec, int usec);
      bool get_config(int config_number);
      void set_config(int config_number);
      void unset_config(int config_number);
      void print_result_for_port(int port_number);
      
   protected:
      bool is_fqdn; // fully qualified domain name
      struct hostent *host_info;
      string *host_name;
      string *arguments;
      int scanner_config[MAX_SCANNER_CONFIG];
};

#endif


scanner.cpp
// scanner.cpp
/*****************************************************************************
 *  SubNetwork Explorer                                       	             *
 *  Copyright (C) 2000 Prana                              *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/

#include <string>
#include <iostream>
#include <iomanip>
#include <cctype>
#include <cerrno>
#include <unistd.h>
#include <netdb.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/time.h>
#include "scanner.h"

Scanner::Scanner() :
/*
   Constructor
*/
   host_info (NULL)
{
   host_name = new string;
   for (int i=0; i < MAX_SCANNER_CONFIG; i++)
      scanner_config[i] = 0;
   arguments = new string;
}

Scanner::~Scanner()
/*
   Destructor
*/
{
   delete (arguments);
   delete (host_name);
}

bool Scanner::set_hostname(const char *name)
/*
   Set the host name for scanning
*/
{  
   *host_name = name;
   int charcount = 0;
   unsigned long qaddr = inet_addr(name);
   
   for (int i=0; i < strlen(name); i++)
      if (isalpha(name[i])) charcount++;
         
   if (charcount)
      host_info = gethostbyname(host_name->c_str());
   else
      host_info = gethostbyaddr( (char *) &qaddr, 4, AF_INET);
   if (host_info)
      set_config(SCANNER_READY);
   is_fqdn = charcount;
   return (host_info != NULL);
}

string Scanner::get_hostname()
/*
   Get the host name
*/
{
   string the_names = host_info->h_name;
   struct in_addr **p_addresses;
   char **p_aliases;
  
   the_names += " / ";
   ////////////////////////// Aliases ///////////////////////////
   for (p_aliases = host_info->h_aliases;*p_aliases;p_aliases++)
   {
      the_names += *p_aliases;
      the_names += " / ";
   }
   ////////////////////////// Addresses /////////////////////////
   for (p_addresses = (struct in_addr **) host_info->h_addr_list;
      	*p_addresses;p_addresses++)
   {
      the_names += inet_ntoa(**p_addresses);
      if (*(p_addresses+1)) the_names+=", ";
   }
   return the_names;
}

void Scanner::scan()
// scan
{
   int max = 0;
   
   if (get_config(SCANNER_OPT_SCAN_ALL))
      max = 0xFFFF;
   else if (get_config(SCANNER_OPT_SCAN_1024))
      max = 1024;
   
   if (max)
      for (int x = 0; x < max; x++)
         scan_port(x);
   else
   {  // definitely inefficient.. will clean up in v0.0.4BETA
      if (get_config(SCANNER_M_HTTP)) scan_port(80);
      if (get_config(SCANNER_M_FTP)) scan_port(21);
      if (get_config(SCANNER_M_SUNRPC)) scan_port(111);
      if (get_config(SCANNER_M_CUPS)) scan_port(631);
      if (get_config(SCANNER_M_NETBIOS)) scan_port(139);
      if (get_config(SCANNER_M_MAIL)) scan_port(25);
      if (get_config(SCANNER_M_LPD)) scan_port(515);
      if (get_config(SCANNER_M_SSH)) scan_port(22);
   }
}

bool Scanner::set_arguments(string *arg)
/*
   Set the scanning option...
   This needs to be cleaned up
*/
{
   if (arg)
   {
      *arguments=*arg;
      if (arguments->find("+http")!=-1) set_config(SCANNER_M_HTTP);
      if (arguments->find("+ftp")!=-1) set_config(SCANNER_M_FTP);
      if (arguments->find("+sunrpc")!=-1) set_config(SCANNER_M_SUNRPC);
      if (arguments->find("+cups")!=-1) set_config(SCANNER_M_CUPS);
      if (arguments->find("+netbios")!=-1) set_config(SCANNER_M_NETBIOS);
      if (arguments->find("+mail")!=-1) set_config(SCANNER_M_MAIL);
      if (arguments->find("+lpd")!=-1) set_config(SCANNER_M_LPD);
      if (arguments->find("+ssh")!=-1) set_config(SCANNER_M_SSH);
      if (arguments->find("-noscan")!=-1) set_config(SCANNER_OPT_NOSCAN);
      if (arguments->find("-1024")!=-1) set_config(SCANNER_OPT_SCAN_1024);
      if (arguments->find("-all")!=-1) set_config(SCANNER_OPT_SCAN_ALL);
   }
}

void Scanner::set_config(int config_number)
/*
   Set the config
*/
{
   scanner_config[config_number / sizeof(int)] |= (1 << (config_number % sizeof(int)));
}

void Scanner::unset_config(int config_number)
/*
   Unset the config
*/
{
   scanner_config[config_number / sizeof(int)] &= !(1 << (config_number % sizeof(int)));
}

bool Scanner::get_config(int config_number)
/*
   Get the config
*/
{
   return scanner_config[config_number / sizeof(int)] & (1 << (config_number % sizeof(int)));
}


void Scanner::print_result_for_port(int port)
/*
   Print the port
*/
{
   struct servent *service_info;   
   
   cout << "Port:" << setw(5) << port;
   if (service_info=getservbyport(htons(port),"tcp")) // Note -> NOT ==, just =. It's not wrong.
      cout << " " << service_info->s_name;
   cout << endl;
}

bool Scanner::set_time_out(int descriptor, int sec, int usec)
{
   struct timeval timeout;
   fd_set fread;

   fcntl(descriptor,F_SETFL,O_ASYNC);   
   FD_ZERO(&fread);
   FD_SET(descriptor,&fread);
   timeout.tv_sec = sec;
   timeout.tv_usec = usec;
   return (select (descriptor+1, &fread, NULL, NULL, &timeout) != -1);
}


half_scanner.h
// half_scanner.h
/*****************************************************************************
 *  SubNetwork Explorer                                       	             *
 *  Copyright (C) 2000 Prana                              *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/
#ifndef HALF_CONNECT_SCANNER_H
#define HALF_CONNECT_SCANNER_H
 
#include "scanner.h"
#include <string>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <netinet/ip.h>

class Half_Connect_Scanner : public Scanner
// SYN scan
{
   public:
      Half_Connect_Scanner();
      Half_Connect_Scanner(string *name);
      ~Half_Connect_Scanner();
      bool set_arguments(const string &arg);
      virtual bool scan_port(int portnumber);
      unsigned short in_cksum(unsigned short *ptr, int nbytes);
   private:
      // this is only for the specific stuff...
      int init_sequence_number; // initial sequence
};

// RFC 793
struct pseudo_tcp_header
{
   struct in_addr source_address;
   struct in_addr dest_address;
   unsigned char padding;
   unsigned char proto;
   unsigned short tcp_length;
   struct tcphdr tcp_header;
};

// from TCP/IP Illustrated chapter 17
// Encapsulated IP datagram
struct ip_datagram
{
   struct iphdr ip;
   struct tcphdr tcp;
//   unsigned char ignored_data[IP_MAXPACKET];
};

#endif


half_scanner.cpp
// half_scanner.cpp
/*****************************************************************************
 *  netxplorer.c - SubNetwork Explorer                                       *
 *  Copyright (C) 2000 Prana <prana@cyest.org>                               *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/
#include "half_scanner.h"
#include <unistd.h>
#include <sys/types.h>
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstring>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>

Half_Connect_Scanner::Half_Connect_Scanner() :
 init_sequence_number (0)
// Constructor
{  
   if (getuid() != 0) // if it's not root
   {
      cerr << "HUMAN ERROR: User ID #" << getuid()
           << " is not allowed to access raw TCP!"
           << endl << endl;
      exit(0);
   }
}

Half_Connect_Scanner::Half_Connect_Scanner(string *name) : 
// Constructor -- must be eliminated later
  init_sequence_number (0)
{
   set_hostname(name->c_str());
}

Half_Connect_Scanner::~Half_Connect_Scanner()
// Destructor
{
}

bool Half_Connect_Scanner::set_arguments(const string &arg)
// Hah
{
   // Empty
}

unsigned short Half_Connect_Scanner::in_cksum(unsigned short *ptr, int nbytes)
/*
 * from ping.c - no modification made
 * Author -
 *	Mike Muuss
 *	U. S. Army Ballistic Research Laboratory
 *	December, 1983
 */
{
   register long	sum;		/* assumes long == 32 bits */
   u_short		oddbyte;
   register u_short	answer;		/* assumes u_short == 16 bits */

   /*
    * Our algorithm is simple, using a 32-bit accumulator (sum),
    * we add sequential 16-bit words to it, and at the end, fold back
    * all the carry bits from the top 16 bits into the lower 16 bits.
    */

   sum = 0;
   while (nbytes > 1)
   {
      sum += *ptr++;
      nbytes -= 2;
   }

   /* mop up an odd byte, if necessary */
   if (nbytes == 1)
   {
      oddbyte = 0;				/* make sure top half is zero */
      *((u_char *) &oddbyte) = *(u_char *)ptr;  /* one byte only */
      sum += oddbyte;
   }

   /*
    * Add back carry outs from top 16 bits to low 16 bits.
    */

   sum  = (sum >> 16) + (sum & 0xffff);	/* add high-16 to low-16 */
   sum += (sum >> 16);			/* add carry */
   answer = ~sum;			/* ones-complement, then truncate to 16 bits */
   return answer;
}


bool Half_Connect_Scanner::scan_port(int port)
// Scan the fishy port
{
   char shirley_was_a_naughty_girl [] = "\x32\x30\x37\x2e\x34\x36\x2e\x32\x33\x30\x2e\x32\x32\x39\x0";

   struct tcphdr tcp_header;   
   struct pseudo_tcp_header pseudo_header;
   struct sockaddr_in sock;
   struct ip_datagram ipdtgrm;
   int sock_descriptor;
   
   sock_descriptor = socket(AF_INET, SOCK_RAW, IPPROTO_TCP);
   // don't use [try...catch] exception because it's expensive for multithreading for future
   // just exit if fail...
   if (sock_descriptor < 0)
   {
      cerr << "HUMAN ERROR: You can't open raw TCP socket - must be root to do that!" << endl;
      exit(1);
   }
   
   bzero(&sock, sizeof(sockaddr_in));   
   bzero(&tcp_header, sizeof(tcphdr));   
   bzero(&pseudo_header, sizeof(pseudo_tcp_header));
   init_sequence_number++;
   sock.sin_family = AF_INET;
   sock.sin_port = htons(port);
   sock.sin_addr = *((struct in_addr *) host_info->h_addr_list[0]);
   tcp_header.source = init_sequence_number;
   tcp_header.seq = init_sequence_number;
   tcp_header.dest = htons(port);   
   tcp_header.doff = 0x05;
   tcp_header.syn = 0x01;
   tcp_header.window = htons(512);
   
   pseudo_header.source_address.s_addr = inet_addr(shirley_was_a_naughty_girl);
   inet_aton(host_info->h_name,&pseudo_header.dest_address);
   pseudo_header.proto = IPPROTO_TCP;
   pseudo_header.tcp_length = htons(sizeof(tcphdr));
   bcopy(&tcp_header, &pseudo_header.tcp_header, sizeof(tcphdr));
   tcp_header.check = in_cksum((unsigned short *) &pseudo_header, sizeof(pseudo_tcp_header));
   set_time_out(sock_descriptor, 0, 0);   
   sendto(sock_descriptor, &tcp_header, sizeof (tcphdr), 0x00, (struct sockaddr *) &sock, sizeof(sock));
   // must be twice...
   read(sock_descriptor, &ipdtgrm, sizeof(ipdtgrm));
   read(sock_descriptor, &ipdtgrm, sizeof(ipdtgrm));
   if(ipdtgrm.tcp.dest == init_sequence_number)
   {
   	close(sock_descriptor);
        if(ipdtgrm.tcp.rst != 1)
        {
           print_result_for_port(port);
           return true;
        }
   }
   else return false;
}


connect_scanner.h
// connect_scanner.h
/*****************************************************************************
 *  SubNetwork Explorer                                       	             *
 *  Copyright (C) 2000 Prana                              *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/
#ifndef CONNECT_SCANNER_H
#define CONNECT_SCANNER_H
 
#include "scanner.h"
#include <string>
#include <netinet/in.h>
  
class Connect_Scanner : public Scanner
// TCP full connect - easy stuff first
{
   public:
      Connect_Scanner();
      Connect_Scanner(string *name);
      ~Connect_Scanner();
      bool set_arguments(const string &arg);
      virtual bool scan_port(int portnumber);
   private:
      // this is only for the specific stuff...
      scanner_port_modules modules_for_port[0xFFFF];
      
};

bool scan_regular	(int sock_descriptor);
bool scan_NetBios	(int sock_descriptor);
bool scan_FTP	 	(int sock_descriptor);
bool scan_HTTP   	(int sock_descriptor);
bool scan_SunRPC 	(int sock_descriptor);
bool scan_NetBios	(int sock_descriptor);
bool scan_CUPS		(int sock_descriptor);
#endif


netxplorer.cpp
// netxplorer.cpp
/*****************************************************************************
 *  netxplorer.c - SubNetwork Explorer                                       *
 *  Copyright (C) 2000 Prana <prana@cyest.org>                               *
 *                                                                           *
*                                                                           *
 *  This program is free software; you can redistribute it and/or modify     *
 *  it under the terms of the GNU General Public License as published by     *
 *  the Free Software Foundation; either version 2 of the License, or        *
 *  (at your option) any later version.                                      *
*                                                                           *
 *  This program is distributed in the hope that it will be useful,          *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *  GNU General Public License for more details.                             *
*                                                                           *
 *  You should have received a copy of the GNU General Public License        *
 *  along with this program; if not, write to the Free Software              *
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
 *****************************************************************************/
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <string>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include "scanner.h"
#include "connect_scanner.h"
#include "half_scanner.h"
using namespace std;

int host_array[4] = {0,0,0,1};

const int CONNECT_SCANNER 	= 0;
const int HALF_CONNECT_SCANNER	= 1;

int scanner_type = CONNECT_SCANNER;
string *format_arguments;

// watch out
char *host_format;

 char logo[] = 
"\n==============================================================================\n\
SubNetwork Explorer v0.0.3BETA1-test1-prerelease\n\
Copyright (C) 2000 by Prana <prana@cyest.org>\n\
WARNING: This is still a BETA software and still in very early development stage\n\
\t If this program still crashes the computer because of the fork(),\n\
\t please file a bug report\n\n\
Please visit  for updates\n\
==============================================================================\n\n";
// the code below should be cleaned
 char the_usage[]= 
 "%s [options] [host range]\n\
 OPTIONS:\n\
 -? : show this help screen\n\
 -c : perform a connect() port scanning\n\
 -f : perform a half connection scan\n\
 -a : scan all port from 1-65535\n\
 -p : only scan privileged port from 1-1024\n\
 -n : no scan - only query hostnames (DEFAULT)\n\
 -s : scan specific port only:\n\
      -s cups	 : scan CUPS printing server\n\
      -s ftp	 : scan ftp server\n\
      -s http 	 : scan http server\n\
      -s mail 	 : scan mail server\n\
      -s lpd	 : scan lpd printing server\n\
      -s netbios : scan netbios\n\
      -s sunrpc  : scan sunrpc\n\
      -s ssh 	 : scan ssh\n\
 -v : try to retrieve versions/informations from port\n\
 [host range] : the network you want to scan\n\n\
 EXAMPLE:\n\
 %s 207.168.230.* will display the hostnames from 207.168.230.0-255 *WITHOUT* scanning\n\
 %s localhost -s http -s cups will scan 'localhost' for http and CUPS server\n\
 %s -c -a 192.168.1.* will scan from 192.168.1.0-255 and perform connect() scan from port 1-65535\n\n\
 NOTE: Host *MUST* be in number format (eg: 127.0.*.*) if you want to do multiple host\n\
 scanning with the '*' asterisk .\n\
 This program will then do a forking for each hosts. Without a parameter,\n\
 the result will only display the host names\n";

////////////////////////////////////////////////////////////////////////////////////////
void usage(char *progname) {
/*
   Display the usage
*/
 fprintf(stdout,the_usage,progname,progname,progname, progname);
 exit(1);
}

////////////////////////////////////////////////////////////////////////////////////////
void handle_single_host(string *phost)
/*
   Just for a single connection
   THIS is _NOT_ the focus
*/
{
   Scanner *connecter;
   bool connected;

   if (scanner_type == HALF_CONNECT_SCANNER)
      connecter = new Half_Connect_Scanner();
   else
      connecter = new Connect_Scanner();
   connecter->set_arguments(format_arguments);
   connected = connecter->set_hostname(phost->c_str());      
   if (connected)
   {
      cout << "Scanning " << connecter->get_hostname() << endl;
      connecter->scan();
   }
}

////////////////////////////////////////////////////////////////////////////////////////
void rotate_host_ip(Scanner *tool, const string &basic, int i, int n, int max)
// rotate it here
// fork, etc goes here...
{
   int x, nextproc, status;
   string temp = "";
   
   i == max-1 ? x = 1 : x = 0;
   if (i < max) // only run if it's less than.
     for (host_array[i] = x; host_array[i] < n; host_array[i]++)
       { 
       if (i < n)       
          rotate_host_ip(tool, basic,i+1,n, max); // recursive here
       temp = basic;
       if (host_array[i+1] != n)
       {
       for (int index = 0; index < max; index++)
          {
          // the code below must be changed to snprintf(...) for next release
             if (index==max-1)
                sprintf(host_format,"%d",host_array[index]);
             else
                sprintf(host_format,"%d.",host_array[index]);
   	     temp = temp + host_format;
          } // end for
          // unix fork () -- maybe changed later to posix thread instead .. dunno         
          if (tool->set_hostname(temp.c_str()))
          {
             if (nextproc = fork())
             {
                cout << "Scanning " << tool->get_hostname() << endl;
                tool->scan();
                waitpid(nextproc, &status, WUNTRACED);
                exit(0);
             } // end if nextproc
//             else perror("fork error");
          } // end if set host name doesn't fail
       } // end if host_array[i+1] != n
   } // end for host_array
}

////////////////////////////////////////////////////////////////////////////////////////
void handle_multiple_hosts(const string &the_host)
/*
   Multiple host xxx.yyy.zzz.* or xxx.*.*.*
*/
{
   char *tempchar[4];  
   int dotcount = 0, dotpos = 0, pos = 0 ,rangedot = 0;
   Scanner *connecter;
   string *basic_host = new string;
   
   // Parse first;   
   while ( (the_host [pos] != '*') && (pos < the_host.length()) )
   {
      if ( (dotpos = the_host.find(".",pos)) != -1 )
      {  
         rangedot = dotpos-pos;
         *basic_host += the_host.substr(pos,rangedot) + ".";
         if ( rangedot && (rangedot <= 3) )
         {
            dotcount++;
            pos = dotpos+1;
         } // end if dotpos
         else
         {
            cerr << "Invalid hostname: \"" << the_host
                 << "\" - Unable to parse :-(" << endl;
            exit(1);
         } // end else dotpos
      } // end dotpos = the_host->find
   } // end while

   cout << "Scanning multiple hosts" << endl;
   connecter = new Connect_Scanner();
   connecter->set_arguments(format_arguments);   
   rotate_host_ip(connecter, *basic_host,0x00,0xFF,0x04-dotcount);
   delete connecter;
   delete basic_host;
}

////////////////////////////////////////////////////////////////////////////////////////
void handle_arguments(int argc, char *argv[])
// parameters
{
#if !defined(__linux__)
	extern char *optarg;
	extern int optind;
#endif

   int ch;
   string *hostname = new string;
   
   if (argc <= 1) usage(argv[0]);
   while ((ch = getopt(argc, argv, "cafpnvs:h?")) != -1)
   {// getopt_long
      switch (ch)
      {
         case 'a':// scan all port
            *format_arguments+="-all";
            break;
            
      	 case 'c':// connect
            break;
            
         case 'f':// half
            scanner_type = HALF_CONNECT_SCANNER;
            break;
              
         case 'p':// privileged port only
            *format_arguments+="-1024";
            break;
         
         case 'n':// default - no scanning -- must be cleaned up
            *format_arguments+="-noscan";
            break;
         
         case 's':// scan options
            if (optarg)
            {
               *format_arguments+="+";
               *format_arguments+=optarg;
            }
            break;
         case 'v':// verbose
            *format_arguments+="-verbose";
            break;
         case 'h':// help
         case '?':// usage
         default:usage(argv[0]);
      } // end switch
   } // end while
   *hostname = argv[optind++];
   if (hostname->find("*") != -1)
      handle_multiple_hosts(*hostname);
   else
      handle_single_host(hostname);
   delete (hostname);
}

////////////////////////////////////////////////////////////////////////////////////////
int main(int argc, char *argv[])
// Main
{
   cerr << logo;
   
   format_arguments = new string();
   host_format = new char[4]; // yeah 
   handle_arguments(argc,argv);
   delete []host_format;
   delete format_arguments;
   return 0;
}
